File: src/main/scala/com/teambrmodding/neotech/common/container/machines/ContainerAbstractMachine.scala
package com.teambrmodding.neotech.common.container.machines

import java.awt.Color

import com.teambrmodding.neotech.common.tiles.AbstractMachine
import com.teambr.bookshelf.common.container.BaseContainer
import com.teambr.bookshelf.common.container.slots.{ICustomSlot, SLOT_SIZE}
import com.teambr.bookshelf.common.tiles.traits.Inventory
import net.minecraft.entity.player.InventoryPlayer
import net.minecraft.item.ItemStack

import scala.collection.mutable.ArrayBuffer

/**
  * This file was created for NeoTech
  *
  * NeoTech is licensed under the
  * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
  * http://creativecommons.org/licenses/by-nc-sa/4.0/
  *
  * @author Paul Davis <pauljoda>
  * @since 1/31/2016
  */
class ContainerAbstractMachine(playerInventory: InventoryPlayer, tile: AbstractMachine) extends
        BaseContainer(playerInventory, tile) {

    lazy val upgradeSlots = new ArrayBuffer[UpgradeSlot]()

    class UpgradeSlot(inventory : Inventory, id : Int, x : Int, y : Int)
            extends RestrictedSlot(inventory, id, x, y) with ICustomSlot {
        override def getSlotSize: SLOT_SIZE.Value = SLOT_SIZE.STANDARD
        override def getPoint: (Integer, Integer) = (xDisplayPosition - 1, yDisplayPosition - 1)
        override def hasColor = true
        override def getColor : Color = new Color(0, 255, 0)
    }

    for(z <- 0 until tile.upgradeInventory.inventoryContents.size()) {
        val slot = new UpgradeSlot(tile.upgradeInventory, z, -10000, -1000)
        upgradeSlots += slot
        addSlotToContainer(slot)
    }

    override def getInventorySizeNotPlayer : Int = tile.getSizeInventory + 6
}

File: src/main/java/com/teambrmodding/neotech/common/container/machines/ContainerAbstractMachine.java
package com.teambrmodding.neotech.common.container.machines;

import com.teambr.bookshelf.common.container.BaseContainer;
import com.teambrmodding.neotech.common.tiles.AbstractMachine;
import com.teambrmodding.neotech.managers.CapabilityLoadManager;
import net.minecraft.entity.player.InventoryPlayer;
import net.minecraft.item.ItemStack;
import net.minecraftforge.items.SlotItemHandler;

import java.util.ArrayList;
import java.util.List;

/**
 * This file was created for NeoTech
 *
 * NeoTech is licensed under the
 * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
 * http://creativecommons.org/licenses/by-nc-sa/4.0/
 *
 * @author Paul Davis - pauljoda
 * @since 2/15/2017
 */
public class ContainerAbstractMachine extends BaseContainer {
    protected AbstractMachine abstractMachine;

    // List of upgrade slots
    public List<SlotItemHandler> upgradeSlots;

    /**
     * Creates the object
     *
     * @param playerInventory The players inventory
     * @param inventory       The tile/object inventory
     */
    public ContainerAbstractMachine(InventoryPlayer playerInventory, AbstractMachine inventory) {
        super(playerInventory, inventory);
        abstractMachine = inventory;

        // Manage upgrade slots
        upgradeSlots = new ArrayList<>();
        for(int id = 0; id < abstractMachine.upgradeInventory.inventoryContents.size(); id++) {
            SlotItemHandler slot = new SlotItemHandler(abstractMachine.upgradeInventory, id, -1000, -1000) {
                /**
                 * Check if the stack is a valid item for this slot. Always true beside for the armor slots.
                 *
                 * @param stack
                 */
                @Override
                public boolean isItemValid(ItemStack stack) {
                    return !abstractMachine.hasUpgradeAlready(stack) &&
                            stack.hasCapability(CapabilityLoadManager.UPGRADE_ITEM_CAPABILITY, null);
                }
            };
            upgradeSlots.add(slot);
            addSlotToContainer(slot);
        }
    }

    /*******************************************************************************************************************
     * BaseContainer                                                                                                   *
     *******************************************************************************************************************/

    /**
     * Get the size of the inventory that isn't the players
     *
     * @return The inventory size that doesn't count the player inventory
     */
    @Override
    public int getInventorySizeNotPlayer() {
        return abstractMachine.getInitialSize() + abstractMachine.upgradeInventory.getSlots();
    }
}
